src/gui/qgsmaptoolcapture.cpp

```diff
@@ -396,7 +397,19 @@ int QgsMapToolCapture::fetchLayerPoint( const QgsPointLocator::Match &match, Qgs
       if ( !f.geometry().vertexIdFromVertexNr( match.vertexIndex(), vId ) )
         return 2;
 
-      layerPoint = f.geometry().constGet()->vertexAt( vId );
+      if ( QgsProject::instance()->topologicalEditing() && match.hasEdge() )
+      {
+        QgsVertexId vId2;
+        if ( !f.geometry().vertexIdFromVertexNr( match.vertexIndex() + 1, vId2 ) )
+          return 2;
+        QgsLineString line( f.geometry().constGet()->vertexAt( vId ), f.geometry().constGet()->vertexAt( vId2 ) );
+
+        layerPoint = QgsGeometryUtils::closestPoint( line,  QgsPoint( match.point() ) );
+      }
+      else
+      {
+        layerPoint = f.geometry().constGet()->vertexAt( vId );
```

Comment:
That's risky, it's using the pointer of a temporary of a shared object. Take a copy of geometry at line 399.